### PR TITLE
Add rx into the app

### PR DESCRIPTION
**Pull request by VitalySokolov**
Changed search box behavior.
Added subscription in header component to user data to show actual info.
Added loading spinner component to prevent user from actions while server is requested.

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -1,14 +1,35 @@
...
+import { LoaderComponent } from '@core/loader/loader.component';
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.css']
 })
-export class AppComponent {
+export class AppComponent implements OnInit {
+  constructor(public userService: UserService, private loaderService: LoaderService, private dialog: MatDialog) {
```
Comment (vitaliy-bobrov) [inline src/app/app.component.ts]: Improve formatting

File: src/app/app.component.ts [hunk 2]
```diff
@@ -1,14 +1,35 @@
...
+  }
+
+  ngOnInit(): void {
```
Comment (vitaliy-bobrov) [inline src/app/app.component.ts]: To be consistent think to use `void` everywhere or never ;)

File: src/app/core/loader/loader.component.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'app-loader',
+  templateUrl: './loader.component.html',
+  styleUrls: ['./loader.component.css']
```
Comment (vitaliy-bobrov) [inline src/app/core/loader/loader.component.ts]: use `OnPush` change detection

File: src/app/shared/loader.service.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+export class LoaderService {
+
+  private loaderState = new Subject<LoaderState>();
+
+  public getLoaderState() {
+    return this.loaderState;
+  }
+
+  public startLoading() {
+    this.loaderState.next({loading: true});
```
Comment (vitaliy-bobrov) [inline src/app/shared/loader.service.ts]: Could be simplified to dispatch just boolean instead object

File: src/app/shared/loading.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { HttpEvent, HttpHandler, HttpInterceptor, HttpRequest, HttpResponse } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { delay, tap } from 'rxjs/operators';
+import { LoaderService } from '@shared/loader.service';
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class LoadingInterceptor implements HttpInterceptor {
+
+  constructor(public loaderService: LoaderService) {
```
Comment (vitaliy-bobrov) [inline src/app/shared/loading.interceptor.ts]: Why `public`?

File: src/app/video-courses/search-course/search-course.component.ts [hunk 6]
```diff
@@ -1,21 +1,19 @@
...
 
-  search = '';
-
-  constructor(private router: Router) {
+  constructor(private courseService: VideoCourseService, private router: Router) {
   }
 
-  onSearchClick() {
-    this.searchCourse.emit(this.search);
+  search(searchString: string) {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/search-course/search-course.component.ts]: Possible to be `undefined`

File: src/app/app.component.ts (hunk 1 above)
Comment (VitalySokolov) [reply src/app/app.component.ts]: Done.

File: src/app/app.component.ts (hunk 2 above)
Comment (VitalySokolov) [reply src/app/app.component.ts]: Done.

File: src/app/shared/loading.interceptor.ts (hunk 5 above)
Comment (VitalySokolov) [reply src/app/shared/loading.interceptor.ts]: It's a leftover, thank you for noticing. I tried to understand why it didn't work.

File: src/app/core/loader/loader.component.ts (hunk 3 above)
Comment (VitalySokolov) [reply src/app/core/loader/loader.component.ts]: Done.

File: src/app/shared/loader.service.ts (hunk 4 above)
Comment (VitalySokolov) [reply src/app/shared/loader.service.ts]: Yes, you are right. Thank you.

**Pull request merged**